modules/carodaBidAdapter.js

```diff
@@ -0,0 +1,212 @@
+      'net';
+    const test = getFirstWithKey(validBidRequests, 'params.test');
+    const currency = getConfig('currency.adServerCurrency');
+    const eids = getFirstWithKey(validBidRequests, 'userIdAsEids');
+    const schain = getFirstWithKey(validBidRequests, 'schain');
+    const request = {
+      auctionId: bidderRequest.auctionId,
+      currency,
+      hb_version: '$prebid.version$',
+      ...ortbCommon,
+      price_type: priceType
+    };
+    if (test) {
+      request.test = 1;
+    }
+    if (schain) {
+      request.schain = schain;
+    }
+    if (config.getConfig('coppa')) {
+      deepSetValue(request, 'privacy.coppa', 1);
+    }
+    if (deepAccess(bidderRequest, 'gdprConsent.gdprApplies') !== undefined) {
+      deepSetValue(
+        request,
+        'privacy.gdpr_consent',
+        bidderRequest.gdprConsent.consentString
+      );
+      deepSetValue(
+        request,
+        'privacy.gdpr',
+        bidderRequest.gdprConsent.gdprApplies & 1
+      );
+    }
+    if (bidderRequest.uspConsent) {
+      deepSetValue(request, 'privacy.us_privacy', bidderRequest.uspConsent);
+    }
+    if (eids) {
+      deepSetValue(request, 'user.eids', eids);
+    }
+    return getImps(validBidRequests, request).map(imp => ({
+      method: 'POST',
+      url: 'https://prebid.caroda.io/api/hb?entry_id=' + pageViewId,
+      data: JSON.stringify(imp)
+    }));
+  },
+  interpretResponse: (serverResponse) => {
+    if (!serverResponse.body) {
+      return;
+    }
+    const { ok, error } = serverResponse.body
+    if (error) {
+      logError(BIDDER_CODE, ': server caught', error.message);
+      return;
+    }
+    try {
+      return JSON.parse(ok.value)
+        .map((bid) => {
+          const ret = {
+            requestId: bid.bid_id,
+            cpm: bid.cpm,
+            creativeId: bid.creative_id,
+            ttl: 300,
+            netRevenue: true,
+            currency: bid.currency,
+            width: bid.w,
+            height: bid.h,
+            meta: {
+              advertiserDomains: bid.adomain || []
+            },
+            ad: bid.ad,
+            placementId: bid.placement_id
+          }
+          if (bid.adserver_targeting) {
+            ret.adserverTargeting = bid.adserver_targeting
+          }
+          return ret
+        })
+        .filter(Boolean);
+    } catch (e) {
+      logError(BIDDER_CODE, ': caught', e);
+    }
+  }
+}
+
+registerBidder(spec)
+
+function getFirstWithKey (collection, key) {
+  for (let i = 0, result; i < collection.length; i++) {
+    result = deepAccess(collection[i], key);
+    if (result) {
+      return result;
+    }
+  }
+}
+
+function getTopUsableWindow () {
+  let res = window;
+  try {
+    while (window.top !== res && res.parent.location.href.length) {
+      res = res.parent;
+    }
+  } catch (e) {}
+  return res;
+}
+
+function getORTBCommon (bidderRequest) {
+  let app, site;
+  const commonFpd = bidderRequest.ortb2 || {};
```

Comment:
First party data must be read from [getConfig('ortb2');](https://docs.prebid.org/dev-docs/publisher-api-reference/setConfig.html#setConfig-fpd)

Reply:
@lksharma according to [src/config.js:l322](https://github.com/prebid/Prebid.js/blob/63f9b0d94292a02100842a50299cbfbb052a08b6/src/config.js#L322), getConfig should never be used to get ortb2. The overwhelming majority of prebid adapters use the bid requests to get this information. Am I missing something obvious?